### ISSUE TITLE
ssr  typescript vue-class-component error

### ISSUE THREAD

**Issue by wjtGitHub** [vuejs/vue-class-component#215]
```typescript
import Vue from "vue";

// 导出一个工厂函数，用于创建新的
// 应用程序、router 和 store 实例
export function createApp() {
  const app = new Vue({
    // 根实例简单的渲染应用程序组件。
    // render: h => h(Example)
    data: {
      url: ""
    },
    template: `<div>访问的 URL 是： {{ url }}</div>`
  });
  return { app };
}
```
webpack
```json
module.exports = {
    resolve: {
        extensions: ['.ts', '.js', '.vue', '.styl', '.json'],
        alias: {
            'vue$': 'vue/dist/vue.esm.js',
        }
    },
}
```

error
![image](https://user-images.githubusercontent.com/17404628/34862342-f17a0e50-f7a4-11e7-81c7-333f20e6dc0c.png)

**Comment by wjtGitHub** [comment]
I use alias 'vue/dist/vue.esm.js'. @

**Comment by wjtGitHub** [comment]
@ktsn

**Comment by ktsn** [comment]
Can you provide a self contained reproduction to confirm?

**Comment by wjtGitHub** [comment]
I can send project.zip file to your email? @ktsn

**Comment by HerringtonDarkholme** [comment]
You also need to use webpack to bundle your server side code, for now.

https://github.com/vuejs/vue/issues/7392

**Comment by wjtGitHub** [comment]
I can provide project's address to you @ktsn

**Comment by wjtGitHub** [comment]
@ktsn https://github.com/wjtGitHub/vue-ssr

**Comment by wjtGitHub** [comment]
@HerringtonDarkholme ,thank you . I have three files that `webpack.config.js`,`server.js`,`client.js` use webpack.

**Comment by HerringtonDarkholme** [comment]
Move 

```ts
module.exports = {
    resolve: {
        extensions: ['.ts', '.js', '.vue', '.styl', '.json'],
        alias: {
            'vue$': 'vue/dist/vue.esm.js',
        }
    },
}
```

to client.js should be fine

**Comment by wjtGitHub** [comment]
@HerringtonDarkholme  
![image](https://user-images.githubusercontent.com/17404628/34864046-5e8bf258-f7ad-11e7-8a48-a6206e6036db.png)

The problem is still in

webpack.config.js
```js
const webpack = require('webpack');
const path = require('path');


module.exports = {
    entry: path.resolve(__dirname, '../src/entry-client.ts'),
    output: {
        path: path.resolve(__dirname, '../dist'),
        filename: 'build.[hash].js',
    },
    module: {
        rules: [{
                test: /\.ts$/,
                exclude: /node_modules/,
                enforce: 'pre',
                loader: 'tslint-loader'
            },
            {
                test: /\.ts$/,
                exclude: /node_modules|vue\/src/,
                loader: 'ts-loader',
                options: {
                    appendTsSuffixTo: [/\.vue$/]
                }
            },
            {
                test: /\.html$/,
                loader: 'raw-loader',
                exclude: [path.resolve(__dirname, '../index.html')]
            },
            {
                test: /\.(gif|jpg|png|woff|svg|eot|ttf)\??.*$/,
                loader: 'url-loader?limit=8192&name=[path][name].[ext]'
            },
            {
                test: /\.styl$/,
                use: [{
                        loader: 'style-loader'
                    }, {
                        loader: 'css-loader'
                    }, {
                        loader: 'postcss-loader',
                        options: {
                            sourceMap: true,
                            plugins: () => [autoprefixer]
                        }
                    },
                    {
                        loader: 'stylus-loader',
                    }
                ]
            },
            {
                test: /\.css$/,
                use: [{
                        loader: 'style-loader'
                    },
                    {
                        loader: 'css-loader'
                    },
                    {
                        loader: 'postcss-loader',
                        options: {
                            plugins: () => [autoprefixer]
                        }
                    }
                ],
                exclude: /\.m\.css$/
            },
            {
                test: /\.m\.css$/,
                use: [{
                        loader: 'style-loader'
                    },
                    {
                        loader: 'css-loader',
                        options: {
                            modules: true,
                            importLoaders: 1,
                            localIdentName: '[path]___[name]__[local]___[hash:base64:5]'
                        }
                    },
                    {
                        loader: 'postcss-loader',
                        options: {
                            plugins: () => [autoprefixer]
                        }
                    }
                ],
            }
        ]
    },
    devtool: 'source-map'
}
```
client.json
```js
const webpack = require('webpack')
const merge = require('webpack-merge')
const baseConfig = require('./webpack.config.js')
const VueSSRClientPlugin = require('vue-server-renderer/client-plugin')
const path = require('path');
const CleanWebpackPlugin = require('clean-webpack-plugin');


module.exports = merge.strategy({
    plugins: 'prepend',
    entry: 'replace',
    output: 'replace'
})(baseConfig, {
    entry: path.resolve(__dirname, '../src/entry-client.ts'),
    output: {
        path: path.resolve(__dirname, '../client'),
        filename: '[name].[hash].js',
    },
    resolve: {
        extensions: ['.ts', '.js', '.vue', '.styl', '.json'],
        alias: {
            'vue$': 'vue/dist/vue.esm.js',
        }
    },
    plugins: [
        new CleanWebpackPlugin([
            'client',
        ], {
            root: path.resolve(__dirname, '../'),
            verbose: false,
            watch: true
        }),
        // 重要信息：这将 webpack 运行时分离到一个引导 chunk 中，
        // 以便可以在之后正确注入异步 chunk。
        // 这也为你的 应用程序/vendor 代码提供了更好的缓存。
        new webpack.optimize.CommonsChunkPlugin({
            name: "manifest",
            minChunks: Infinity
        }),
        // 此插件在输出目录中
        // 生成 `vue-ssr-client-manifest.json`。
        new VueSSRClientPlugin()
    ]
})
```

server.js
```js
const merge = require('webpack-merge')
const nodeExternals = require('webpack-node-externals')
const baseConfig = require('./webpack.config.js')
const VueSSRServerPlugin = require('vue-server-renderer/server-plugin')
const path = require('path');
const CleanWebpackPlugin = require('clean-webpack-plugin');

module.exports = merge.strategy({
    plugins: 'prepend',
    entry: 'replace',
    output: 'replace'
})(baseConfig, {
    // 将 entry 指向应用程序的 server entry 文件
    entry: path.resolve(__dirname, '../src/entry-server.ts'),
    // 这允许 webpack 以 Node 适用方式(Node-appropriate fashion)处理动态导入(dynamic import)，
    // 并且还会在编译 Vue 组件时，
    // 告知 `vue-loader` 输送面向服务器代码(server-oriented code)。
    target: 'node',
    // 对 bundle renderer 提供 source map 支持
    devtool: 'source-map',
    // 此处告知 server bundle 使用 Node 风格导出模块(Node-style exports)
    output: {
        libraryTarget: 'commonjs2',
        path: path.resolve(__dirname, '../server'),
        filename: '[name].[hash].js',
    },
    resolve: {
        extensions: ['.ts', '.js', '.vue', '.styl', '.json'],
        alias: {
            'vue$': 'vue/dist/vue.esm.js',
        }
    },
    // https://webpack.js.org/configuration/externals/#function
    // https://github.com/liady/webpack-node-externals
    // 外置化应用程序依赖模块。可以使服务器构建速度更快，
    // 并生成较小的 bundle 文件。
    externals: nodeExternals({
        // 不要外置化 webpack 需要处理的依赖模块。
        // 你可以在这里添加更多的文件类型。例如，未处理 *.vue 原始文件，
        // 你还应该将修改 `global`（例如 polyfill）的依赖模块列入白名单
        whitelist: /\.css$/
    }),
    // 这是将服务器的整个输出
    // 构建为单个 JSON 文件的插件。
    // 默认文件名为 `vue-ssr-server-bundle.json`
    plugins: [
        new CleanWebpackPlugin([
            'server',
        ], {
            root: path.resolve(__dirname, '../'),
            verbose: false,
            watch: true
        }),
        new VueSSRServerPlugin()
    ]
})
```

**Comment by HerringtonDarkholme** [comment]
Remove 

        alias: {
            'vue$': 'vue/dist/vue.esm.js',
        }

in server.js

**Comment by wjtGitHub** [comment]
@HerringtonDarkholme Thank you ,but the problem still in. Can you check the configuration file for me.
address: https://github.com/wjtGitHub/vue-ssr
1.git clone https://github.com/wjtGitHub/vue-ssr
2.npm i
3.npm run client
4.npm run server
5.node src/www.js
thank you very much.You are my hope

**Comment by HerringtonDarkholme** [comment]
Sorry, my mistake.

The culprit is 
```
    target: 'node',
    // 对 bundle renderer 提供 source map 支持
    devtool: 'source-map',
    // 此处告知 server bundle 使用 Node 风格导出模块(Node-style exports)
    output: {
        libraryTarget: 'commonjs2',
        path: path.resolve(__dirname, '../server'),
        filename: '[name].[hash].js',
    },
```

You cannot use typescript as `commonjs` because TS won't compile `import V from 'commonjs'` to `V = require('commonjs')`. 

Please patiently wait TS2.7. Or pipe your TS output to Babel for proper commonjs/esm interop.

TLDR; You cannot easily build a server side code in TS, for now.

**Comment by wjtGitHub** [comment]
Ok , thank you very much.
